middleware/plugin/ngx_http_casual_module.c

```diff
@@ -0,0 +1,688 @@
+   {
+      ngx_del_timer(&context->timer);
+   }
+
+   casual_http_inbound_deallocate( &context->casual_handle);
+}
+
+static void initilize_context( ngx_http_casual_ctx_t* context)
+{
+   context->casual_handle.fd = 0;
+   context->casual_handle.context_holder = NULL;
+}
+
+static ngx_int_t get_header_length( ngx_http_request_t* r)
+{
+   ngx_list_part_t *part = &r->headers_in.headers.part;
+
+   ngx_uint_t length = 0;
+   ngx_uint_t i = 0;
+
+   for( i = 0; /* void */; i++)
+   {
+      if( i >= part->nelts)
+      {
+         if( part->next == NULL)
+         {
+            /* The last part, search is done. */
+            break;
+         }
+
+         part = part->next;
+         i = 0;
+      }
+      length++;
+   }
+   return length;
+}
+
+static void copy_headers( ngx_http_request_t* r, casual_http_inbound_header_t* header)
+{
+   ngx_list_part_t *part;
+   ngx_table_elt_t *h;
+   ngx_uint_t i;
+
+   /*
+    Get the first part of the list. There is usual only one part.
+    */
+   part = &r->headers_in.headers.part;
+   h = part->elts;
+
+   /*
+    Headers list array may consist of more than one part,
+    so loop through all of it
+    */
+   casual_http_inbound_header_t* headerpointer = header;
+   for( i = 0; /* void */; i++)
+   {
+      if( i >= part->nelts)
+      {
+         if( part->next == NULL)
+         {
+            /* The last part, search is done. */
+            break;
+         }
+
+         part = part->next;
+         h = part->elts;
+         i = 0;
+      }
+
+      headerpointer->key = copy_from_ngx_str_t( h[i].key);
+      headerpointer->value = copy_from_ngx_str_t( h[i].value);
+      headerpointer++;
+   }
+}
+
+static void free_headers( casual_http_inbound_header_t* header, ngx_int_t size)
+{
+   for( ngx_int_t i = 0; i < size; ++i)
+   {
+      free( header[i].key);
+      free( header[i].value);
+   }
+   free( header);
+   header = 0L;
+}
+
+static ngx_int_t set_custom_header_in_headers_out( ngx_http_request_t* r, casual_http_inbound_headers_t* header)
+{
+   for( size_t i = 0; i < header->size; i++)
+   {
+      ngx_table_elt_t* h;
+
+      /*
+      All we have to do is just to allocate the header...
+      */
+      h = ngx_list_push( &r->headers_out.headers);
+      if (h == NULL) {
+         return NGX_ERROR;
+      }
+
+      /*
+      ... setup the header key ...
+      */
+      h->key = copy_to_ngx_str_t( header->data[i].key, r);
+      ngx_log_debug1( NGX_LOG_DEBUG_ALL, r->connection->log, 0, "xatmi: key [%V]", &h->key);
+
+
+      /*
+      ... and the value.
+      */
+      h->value = copy_to_ngx_str_t( header->data[i].value, r);
+      ngx_log_debug1( NGX_LOG_DEBUG_ALL, r->connection->log, 0, "xatmi: value [%V]", &h->value);
+
+
+      /*
+      Mark the header as not deleted.
+      */
+      h->hash = 1;
+   }
+    return NGX_OK;
+}
+
+static ngx_int_t buffer_size( ngx_chain_t* response)
+{
+   ngx_int_t result = 0;
+   for( ngx_chain_t *cl = response; cl; cl = cl->next)
```

Comment:
`cl`? Is `current` a better name?

Reply:
True, I pretty much stole that example code from tutorials. :-)
I will review all "nginx plugin code" and make it more human readable. :-)